### PR TITLE
postgresql: Implement automatic extension preloading

### PR DESCRIPTION
**Pull request by Lillecarl**
## Description of changes

Implement `passthru.shared_preload_library` -> `services.postgresql.settings.shared_preload_libraries` with option to disable.
This is good because you don't have to wade through documentation just to enable an extension, making it easier for newcomers to have a really pimped out postgresql. A pimped postgres makes for happy nerds!

So for those reading that don't know what shared_preload_libraries is. Some pg extensions require being loaded when pg starts rather than arbitrarily (global state <3). If they have these requirements I see no reason not to export this information to the end user by automating the process entirely. My implementation is subject to change if a nixos module master swings by.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@markuskowa This is one of those things i said "meta" was underutilized for! 😄

### CODE REVIEW

**Comment by Lillecarl** [comment]
@Ma27 @wolfgangwalther Since I saw you guys in #341421 I'll just notify you and see if you'd like to weigh in. Maybe the PR should be split into the module and the package passthrus separately, or the change might be entirely unwanted. Eyes wouldn't hurt 😄

**Review by wolfgangwalther — COMMENTED** [review]
I don't have an opinion on the modules part, since I only use nixpkgs, not NixOS.

I wondered whether an extension could provide multiple shared preload libraries (and thus a simple string wouldn't be enough) - but when I searched github and tested a random sample, I couldn't find any.

In general, metadata about extension and enabling some kind of "discoverability" is not bad, I guess.

File: pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix [hunk 1]
```diff
@@ -87,6 +87,8 @@ in
   };
 
   passthru = {
+    shared_preload_library = "vectors.so";
```
[the remark this answers is not in the document]
Comment (Lillecarl) [reply pkgs/servers/sql/postgresql/ext/pgvecto-rs/default.nix:90]: https://docs.pgvecto.rs/getting-started/installation.html

Weirdly enough it is correct (as per docs). I'll add a comment :)

**Comment by Lillecarl** [comment]
@danbst
@thoughtpolice

File: nixos/modules/services/databases/postgresql.nix [hunk 2]
```diff
@@ -481,7 +497,29 @@ in
...
+        found_libs =
+          (builtins.map (plugMap: plugMap.passthru.shared_preload_library)
+            (builtins.filter (plugFilter: (plugFilter.passthru.shared_preload_library or null) != null)
+              (cfg.extraPlugins { postgresql = cfg.package; })));
+
+        all_libs =
+          (cfg.settings.shared_preload_libraries or [ ]) ++
+          (if cfg.autoloadPlugins then found_libs else [ ]);
+      in
+      cfg.settings // {
```
[the remark this answers is not in the document]
Comment (Ma27) [reply nixos/modules/services/databases/postgresql.nix:511]: That said, if you need assistance in implementing that, let me know!

**Comment by Lillecarl** [comment]
Thanks all for taking a look, I reconsidered and I think it might be a bit too much magic for no benefit, if you're running Postgres plugins you should know what you're doing with Postgres. (I also became unemployed and lost access to everything but a Chromebook)

**Pull request closed without merging**